horovod/mxnet/mpi_ops.cc

```diff
@@ -0,0 +1,411 @@
+// Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// =============================================================================
+
+#include <chrono>
+#include <memory>
+#include <thread>
+
+#include "../common/operations.h"
+#include "adapter.h"
+#include "cuda_util.h"
+#include "handle_manager.h"
+#include "mpi_ops.h"
+#include "ready_event.h"
+#include "tensor_util.h"
+
+namespace horovod {
+namespace mxnet {
+
+static HandleManager handle_manager;
+
+namespace {
+
+std::string GetOpName(std::string prefix, char* name) {
+  if (name != nullptr) {
+    return prefix + "." + std::string(name);
+  }
+  return std::string();
+}
+
+std::string GetOpNameHandle(std::string prefix, const std::string& name, int handle) {
+  if (name.length() != 0) {
+    return name;
+  }
+  return prefix + ".noname." + std::to_string(handle);
+}
+
+} // namespace
+
+int DoAllreduce(NDArray* tensor, NDArray* output, int average, const std::string& name, Callback cb) {
+  ThrowIfError(common::CheckInitialized());
+ 
+  auto handle = handle_manager.AllocateHandle(cb);
+  auto device = TensorUtil::GetDevice(tensor);
+  auto hvd_tensor = std::make_shared<MXTensor<NDArray>>(tensor);
```

Comment:
You may need ready event here as well in case you do any async CUDA calls in MXNet.  In both PyTorch and TensorFlow we've found that you need to wait for current CUDA threads to finish before data is ready for NCCL and MPI.

Reply:
After discussion, we came to the conclusion that all async CUDA calls in MXNet have a `ctx.get_stream<gpu>()->Wait();` at the end of the call pushing them to MXNet engine. For example when copying data from CPU to GPU asynchronously, we use such a call [here](https://github.com/apache/incubator-mxnet/blob/5a83b6b563211f430688e41eab4752c6de4ecf22/src/ndarray/ndarray.cc#L1245) to ensure data is ready before the next CUDA kernel happens. This should guarantee that the current CUDA threads finish before data is ready for NCCL and MPI.

Reply:
Does it cover a situation where the data is originally on GPU, and no copy is needed?  NCCL CUDA stream is different from the original CUDA stream, so we still need to synchronize on original stream before NCCL stream starts.

Reply:
Yes, `ctx.get_strea<gpu>()->Wait()` is also called when copying from GPU to GPU in MXNet. Please see [here](https://github.com/apache/incubator-mxnet/blob/5a83b6b563211f430688e41eab4752c6de4ecf22/src/ndarray/ndarray.cc#L1257)

Reply:
I think what @apeforest is saying is that even when the data is originally on GPU, it will have been the result of _some_ MXNet operation. My understanding is that for practically all MXNet operations using the Push API, it will have a `ctx.get_stream<gpu>()->Wait()` after the current CUDA threads have finished for that stream. Therefore, we can be assured that the data is ready and synchronized on original stream before NCCL starts.